src/source-odp.c

```diff
@@ -0,0 +1,286 @@
+    }
+
+    PacketDecodeFinalize(tv, dtv, p);
+
+    SCReturnInt(TM_ECODE_OK);
+}
+
+static TmEcode ReceiveODPBreakLoop(ThreadVars *ptv, void *arg)
+{
+    odp_loop_break = 1;
+    SCReturnInt(TM_ECODE_OK);
+}
+
+/**
+ * \brief Registration Function for RecieveODP.
+ * \todo Unit tests are needed for this module.
+ */
+void TmModuleReceiveODPRegister (void)
+{
+    tmm_modules[TMM_RECEIVEPCAP].name = "ReceiveODP";
+    tmm_modules[TMM_RECEIVEPCAP].ThreadInit = ReceiveODPThreadInit;
+    tmm_modules[TMM_RECEIVEPCAP].Func = NULL;
+    tmm_modules[TMM_RECEIVEPCAP].PktAcqLoop = ReceiveODPLoop;
+    tmm_modules[TMM_RECEIVEPCAP].PktAcqBreakLoop = ReceiveODPBreakLoop;
+    tmm_modules[TMM_RECEIVEPCAP].ThreadExitPrintStats = NULL;
+    tmm_modules[TMM_RECEIVEPCAP].ThreadDeinit = NULL;
+    tmm_modules[TMM_RECEIVEPCAP].RegisterTests = NULL;
+    tmm_modules[TMM_RECEIVEPCAP].cap_flags = SC_CAP_NET_RAW;
+    tmm_modules[TMM_RECEIVEPCAP].flags = TM_FLAG_RECEIVE_TM;
+}
+
+static inline Packet *_odp_process_packet(ODPThreadVars *ptv, odp_packet_t opkt)
+{
+
+    uint32_t pkt_len = odp_packet_len(opkt);
+    uint8_t *pkt_data = odp_packet_data(opkt);
+    Packet *p;
+
+    /* make sure we have at least one packet in the packet pool, to prevent
+     * us from alloc'ing packets at line rate */
+    PacketPoolWait();
+
+    p = PacketGetFromQueueOrAlloc();
+    if (p == NULL)
+        return NULL;
+
+    PKT_SET_SRC(p, PKT_SRC_WIRE);
+    gettimeofday(&p->ts, NULL);
+    p->datalink = LINKTYPE_ETHERNET;
+    PacketSetData(p, pkt_data, pkt_len);
+    p->pkt_odp = opkt;
+    p->ReleasePacket = ODPReleasePacket;
+    p->flags |= PKT_IGNORE_CHECKSUM;
+    return p;
+}
+
+/**
+ *  \brief Main ODP reading Loop function
+ */
+TmEcode ReceiveODPLoop(ThreadVars *tv, void *data, void *slot)
+{
+    ODPThreadVars *ptv = (ODPThreadVars *)data;
+    TmSlot *s = (TmSlot *)slot;
+    uint64_t sched_wait = odp_schedule_wait_time(ODP_TIME_SEC_IN_NS * 1);
+    odp_event_t ev;
+    odp_packet_t pkt;
+    Packet *p = NULL;
+    int ret;
+
+    SCEnter();
+    ret = odp_init_local(instance, ODP_THREAD_WORKER);
+    if (ret) {
+        fprintf(stderr, "error in odp_init_local\n");
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    ptv->slot = s->slot_next;
+    ptv->cb_result = TM_ECODE_OK;
+
+    SCLogInfo("ODP worker thread enter to loop");
+    while (!odp_loop_break)
+    {
+        ev = odp_schedule(NULL, sched_wait);
+        if (ev == ODP_EVENT_INVALID)
+		continue;
+
+        pkt = odp_packet_from_event(ev);
+        if (!odp_packet_is_valid(pkt))
+            continue;
+
+        p = _odp_process_packet(ptv, pkt);
+        if (!p) {
+            odp_packet_free(pkt);
+            break;
+        }
+
+        if (TmThreadsSlotProcessPkt(ptv->tv, ptv->slot, p) != TM_ECODE_OK) {
+            TmqhOutputPacketpool(ptv->tv, p);
+            SCReturnInt(TM_ECODE_FAILED);
+        }
+    }
+
+    SCLogInfo("ODP worker thread exited");
+    StatsSyncCountersIfSignalled(tv);
```

Comment:
Here we have a problem, the thread will sync its counter only at exit if I read the code correctly. This means there will be no stats available during a normal run and this will be available only at exit. Code should be updated to regularly call the StastSyncCountersIfSignalled function.